formulaic/transforms/featurehash.py

```diff
@@ -0,0 +1,68 @@
+from hashlib import md5
+from numbers import Number
+from typing import Any, Callable, Dict, Iterable, List, Optional, Union
+
+import numpy as np
+
+from formulaic.materializers.types import FactorValues
+
+from .contrasts import Contrasts, encode_contrasts
+
+
+def md5_to_int(s: str) -> int:
+    return int(md5(s.encode(), usedforsecurity=False).hexdigest(), 16)
+
+
+def hashed(
+    data: Any,
+    levels: int,
+    contrasts: Optional[
+        Union[Contrasts, Dict[str, Iterable[Number]], np.ndarray]
+    ] = None,
+    *,
+    hash_func: Callable[[str], int] = md5_to_int,
+    spans_intercept: bool = False,
+):
+    """
+    Hashes the values of a factor into a fixed number of levels. DUe to the
+    large number of levels required to avoid collisions, this transform should
+    be used with output='sparse' to avoid memory issues.
```

Comment:
note: If you *did* make it a stateful transform you could peek at the model spec and validate this explicitly, posting a warning if `output` is not `'sparse'`.